src/bentoml/models.py

```diff
@@ -318,6 +318,35 @@ def create(
         res.exit_cloudpickle_context(imported_modules)
 
 
+@inject
+@contextmanager
+def save(
+    name: Tag | str,
+    *,
+    module: str = "",
+    api_version: str | None = None,
+    metadata: dict[str, t.Any] | None = None,
+    _model_store: ModelStore = Provide[BentoMLContainer.model_store],
+) -> t.Generator[Model, None, None]:
+    api_version = "v1" if api_version is None else api_version
+    res = Model.create(
+        name,
+        module=module,
+        api_version=api_version,
+        signatures={},
+        metadata=metadata,
+        context=ModelContext(framework_name="", framework_versions={}),
+    )
+
+    try:
+        yield res
+    except Exception:
+        raise
```

Comment:
Leaving an open exception like this is done on purpose. The user code will actually be injected into the try block when calling `with bentoml.models.save():`. So if there is any exception raised by the user's code, we shouldn't handle it differently and just raise it. I believe that is why they did the same thing in `bentoml.models.create`

Reply:
No I don't think this is correct. All exceptions should either wrapped within BentoMLException and raise from it.

Reply:
Ming is correct, this is so that `flush` and `save` aren't run if user code raises an exception. We don't want to mess with the actual exception in this case.

Reply:
catching a general `Exception` and reraise it is just like we don't catch. It isn't necessary at all.

The following examples are equivalent to each other:

```python
try:
    model.save_pretrained(res.path)
    tokenizer.save_pretrained(res.path)
except Exception:
    raise
```

```python
model.save_pretrained(res.path)
tokenizer.save_pretrained(res.path)
```

Reply:
I'm actually not sure this try block should exist, I thought that the context manager would ignore the exception but that isn't true.

Reply:
I'm not really sure how `yield res` could fail, anyway?

Reply:
We definitely don't want to wrap user exceptions in a `BentoMLException`.

Reply:
> I'm not really sure how `yield res` could fail, anyway?

Oh, definitely:
```pytb
In [1]: from contextlib import contextmanager

In [2]: @contextmanager
   ...: def catching():
   ...:     try:
   ...:         yield
   ...:     except Exception as e:
   ...:         print(f"The exception is {type(e)}: {e}")
   ...:         raise
   ...:     finally:
   ...:         print('finally called')
   ...: 

In [3]: with catching():
   ...:     1/0
   ...: 
The exception is <class 'ZeroDivisionError'>: division by zero
finally called
```

Reply:
I think we might want to wrap the users exception in a exceptiongroup stack. https://peps.python.org/pep-0654/

Reply:
@aarnphm could you explain the benefits of using an exception group. And is this feature backward compatible with python 3.8 etc.?